src/Peers/provider.cpp

```diff
@@ -59,80 +61,94 @@ void Provider::listen() {
+            static_pointer_cast<TaskResponse>(followerPayload);
 
-            shared_ptr<TaskResponse> aggregatePayload =
-                make_shared<TaskResponse>(aggregatedResults);
-            Message aggregateResultMsg(uuid, IpAddress(host, port),
-                                       aggregatePayload);
+        // append to followerData
+        followerData.push_back(taskResp->getTrainingData());
+        server->replyToConn("Received follower result.");
+        server->closeConn();
+    }
 
-            cout << "Waiting for connection back to requester" << endl;
-            // busy wait until connection is established
-            while (client->setupConn(reqHost, reqPort, "tcp") == -1) {
-                sleep(5);
-            }
+    cout << endl;
+    workloadThread.join();
+    TaskResponse aggregatedResults = aggregateResults(followerData);
 
-            client->sendRequest(aggregateResultMsg.serialize().c_str());
-        } else {
-            // if this is a follower, send the result to the leader peer
-            cout << "Waiting for connection back to leader" << endl;
-            // get the ip from the bootstrap server
-            // ------------------ hard code for now ------------------
-            IpAddress leaderIp =
-                task->getAssignedWorkers()[task->getLeaderUuid()];
-            const char* leaderHost = leaderIp.host.c_str();
-            const char* leaderPort = to_string(leaderIp.port).c_str();
-            // busy wait until connection is established with the leader
-            while (client->setupConn(leaderHost, leaderPort, "tcp") == -1) {
-                sleep(5);
+    // Send results back to requester
+    // TODO: requester IP address could change
+    shared_ptr<TaskResponse> aggregatePayload =
+        make_shared<TaskResponse>(aggregatedResults);
+    Message aggregateResultMsg(uuid, IpAddress(host, port),
+                                aggregatePayload);
+
+    cout << "Waiting for connection back to requester" << endl;
+    // Keep trying to send results back to requester 
+    while (true) {
+        // busy wait until connection is established
+        while (client->setupConn(requesterIpAddr, "tcp") != 0) {
+            sleep(5);
+        }
+        if (client->sendMsg(aggregateResultMsg.serialize().c_str()) != 0) {
+            continue;
+        }
+
+        if (!server->acceptConn()) continue;
+
+        // receive response from requester
+        string serializedData = server->receiveFromConn();
+        server->closeConn();
+        // process this request
+        Message msg;
+        msg.deserialize(serializedData);
+        shared_ptr<Payload> payload = msg.getPayload();
+
+        if (payload->getType() == Payload::Type::GENERIC_RESPONSE) {
+            shared_ptr<GenericResponse> gr =
+                static_pointer_cast<GenericResponse>(payload);
+            if (gr->getStatus() == GenericResponse::Status::SUCCESS) {
+                break;
             }
```

Comment:
Not needed right now, but it would help to log different branches of execution (ie. what does a Status::FAILED mean  and why we are retrying the loop)

Reply:
I created the generic response payload type as a generic acknowledgement. In this use case, the leader peer keeps trying to send the results back to the requester until the requester receives it and sends back this generic response with success. I guess in this specific case, there doesn't seem to be a scenario where the requester would send back a generic response with Status::FAILED.

A better solution might be to create a separate acknowledgement payload type with no status (or even just assume an empty payload is an acknowledgement). Thoughts?

Reply:
I think that is more clear (ie the message type is `acknowledgement`). 

Not a big deal and I am good to merge this as is.